### PR TITLE
Refactoring: remove mutex from bucketChunkReader

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does
None would like to review a PR removing a mutex. If there's a mutex, it's very likely there's a very good reason. However, I can't find a good reason for `bucketChunkReader.mtx` so in this PR I'm proposing to remove it (please review this carefully).

I think the mutex was there to protect the access from `loadChunks()` to a set of chunks and statistics, but recent refactorings replaces both of them with a concurrency safe version:
- https://github.com/grafana/mimir/pull/3506: in this PR we changed `chunkBytes []*[]byte` with `*pool.BatchBytes` which is concurrency safe
- https://github.com/grafana/mimir/pull/3509: in this PR we changed `*queryStats` with `*safeQueryStats` which is concurrency safe

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/bucket_chunk_reader.go [hunk 1]
```diff
@@ -106,15 +108,8 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesEntry, a
...
-	locked := true
-	r.mtx.Lock()
-
 	defer func() {
 		stats.merge(&localStats)
-
-		if locked {
-			r.mtx.Unlock()
-		}
 	}()
```
Comment (colega) [inline pkg/storegateway/bucket_chunk_reader.go:113]: Nit: `defer stats.merge(&localStats)` instead?

**Review by colega — APPROVED** [review]
I think the change makes sense. 

I rewviewed the calls, and they all seem to be thread safe. Didn't check on the caller part, but if this was called with same indexes, then the results were already concurrently-safe overwritten, which isn't better.

File: pkg/storegateway/bucket_chunk_reader.go (hunk 1 above)
Comment (pracucci) [reply pkg/storegateway/bucket_chunk_reader.go]: Right, done: https://github.com/grafana/mimir/pull/3606/commits/2a16db2c87becf8265d47f31028814188f59eed8

**Comment by pstibrany** [comment]
I looked at this too, and I cannot figure out what was the mutex actually protecting. It seems safe to remove.

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, I think we removed the critical sections in the PRs you have linked in the description

**Pull request merged**